Fixed blocks state <-> meta conversion

shartte opened the pull request:
Fixes #23.

The getMetaFromState and getStateFromMeta methods need to be implemented to convert the block states into a meta value that gets actually written to the savegame for the block.

p.s.: I simply used the enum ordinal here. I toyed with the idea to use a switch, but it seems unlikely they're going to go 4D anytime soon ;-).

shartte committed:
Fixes #23 by implementing getMetaFromState/getStateFromMeta to correctly persist the block state for pillars.

Elix-x commented:
This is the same way vanilla logs/pillars are implemented, right?

shartte commented:
They're using the meta as a bitfield in vanilla, but I don't know why. Since the column can't simultaneously point into the X AND the Y direction for example. 
But yeah, whenever you want a block state to be persisted to the meta of the block (in case you dont have a tile entity), getMetaFromState/getStateFromMeta are needed.

Elix-x commented:
I know about meta<->state methods, i just did not implement them.
They are using bitfield because of variants (aka oak/birch/...), so in this case use of axis is better.
Meanwhile, can you please check all blocks that don't have these methods serialize state and add them to this PR?

shartte committed:
Added meta persistence for quartz torch (based on vanilla torch).

shartte commented:
Okay, I added the functions for the Quartz Torch. That was the only thing I could find that didn't have it yet.
For the controller, I'd like to add it in a separate PR where I also add the actual graphics, since I am almost done with that.

Elix-x commented:
Way you do this with switch is a bit weird? Is it because vanilla does it this way? Why not just use ordinals of facing?

Because i think quartz torch is attachable to the ceilling. @yueh correct?

shartte commented:
Yeah it's a copy of how vanilla does it. If the torch is attachable to the ceiling, I actually need to recheck, you're right.

Elix-x commented:
Because i think it is (if not, it will be nice feature anyway).

shartte commented:
I have a 1.7 client open right now anyway. I'll verify there.
Yeah it can go to the ceiling. I'll just switch to ordinals.

shartte committed:
Allow for quartz fixtures to face downwards as well.

shartte commented:
Allright. I've implemented the downward facing as well. Sadly we can't use BlockTorch.FACING then because that property doesn't allow EnumFacing.DOWN. 
I've added a custom property for the quartz torch that allows all directions.

Elix-x commented:
Totally reasonable.

shartte commented:
Yay! First PR merged :)

Elix-x commented:
<sub><sup><sub><sup>4̶̝͔ͮͪ̑ͬ̐͋ͣ̑ͥ̕ ̧̘̲̬̓̊̇̊̓͡M̒̓̽͗̄͒̚͠͏͚̞O̦͉̻̳̲͙̲ͨ̀ͨ͌͐Ḁ̬̮͇̍͞Ŗ̞͔̲̬̘̈̌͋͆̄͐̑ ͓͈̲̫͚̦̎͛̍̚͠ͅ2̴̰̯̺̮̪̮̦̯ͭ̊̆͌̔̀̇ ̸͍͇̰̪̐̏͟G̛͕ͭͧ̊̈̑̌͊͢O̧͚̠̝̻ͬ͗̔ͩ́ͨ̚͢͝ ͙̩̫̠̲̮̩̠͐͊͛ͨA̺̼̿ͫͧ̊̊͌ͬŃ̞̤ͬ͟ ̵̸̣̻͓͎͔͔̼͛ͤ̈́̔̍̿͛̂U̙͉̺͎͖ͥ̐͐̾͐͗̐̇͂͝ ̨̥̹̦̙̓̅̀̿͘R̢͎̲͕̝̤̭͍ͣ̏̓̒ͨ͐͑̚ ̶̻̹̌̃ͪͪ͠I̡̛͚͉̟̍̏ͅN̴̢̮̹̼͓̭̫̓́͑̈̂̚͘</sup></sub></sup></sub>

yueh commented:
This is a flawed fix.

1) Quartz fixtures have currently 12 rotations. This is mostly due to using metadata, which limits it to 16 at max, otherwise it should have 24 if possible (which it is with 1.8+). 
So currently it misses at least 6 possible rotations.

2) `AEBaseBlock` is the base for every single block of AE. This means adding a property to it does more or less pollute every other block with a property already present (or at least it should be).
Rotating blocks is pretty central in AE. Pretty much everything is rotatable, just the basic blocks use a limited set of 12 orientations by using block metadata instead of a full tile entity. These also apply to blocks like skystone.

Especially the decorative blocks (and probably world gen) are a good candiate to separate mod.
These are very basic blocks usually and the current blockstate pretty much cover any requirements for them, without depending on the full functionality tile entities require.
Also easier to port and at least provide the basic blocks including worldgen for a new version, so people could already start to collect resources and release the rest afterwards without having them to throw the world away.

Elix-x commented:
Why anybody would need to rotate skystone? A torch besides it's 6 placements?
I understand that you may want a torch attached to ceilling facing north and flipped, but we can't. We must NOT use a TE just for rotations whenever possible. TEs decrease performance drastically for 0.01% of people that want to rotate their torch. On top of that, this is not decocraft (i would agree with you if it was), we shouldn't provide complex advanced decoration blocks, it's simply not our job.

yueh commented:
Just look at the current features of AE.

The torches are not symmetrical like vanilla torches, so someone might want to rotate them that all tips face the center of a room or that it faces away from a door. Whatever, just use your imagination.

Skystone uses different rotations to let meteors not look like a simple flat wall and break up the texture a bit. That is something we could achieve with the alternate textures 1.8+ provides, but some players might want to have a flat wall AND randomised meteors. Why limit them?

Normal blockstates (metadata based) are still limited to 16 variants (iirc), so if there are technical limitations preventing us from providing all 24 orientations, it would still be fine to limit it to 12 like currently. (For torches it would actually be 48 to cover everything)

Elix-x commented:
Random textures is just random rendering and has nothing to do with meta. We can specify as many of them as we want. In order to avoid unneeded rotations, we rotate each texture and add all 4 rotational variants.

"Ideally..." we would be decocraft. But we are not. Players use AE for storage/autocrafting, not decorations. And yes, there is 0.1% of them who will be actively using AE blocks instead of chisel's. Now 1% of 0.1% are extreme OCD so they will suicide if torch cannot be rotated on another axis to be perfectly symmetrical.
So yeah, 0.001%... There are times when we should simply stop.

yueh commented:
Random texture is useless, because that is a hard choice. Either you have it or not. So no possibility to have randomised meteors AND a flat wall.

Also being rotatable is nearly an integral part in 1.8+, as block states have a rotate method (and similar).
It is still limited to 16 states, but that will go away as metadata will be phased out at some point. Maybe already with 1.11.

Further it is not decocraft just a convention, that blocks are rotatable. Either all 24 permutations, with TE or less with a normal block (say 12 or 16). The current quartz fixtures allows to be attached to all 6 faces (not really a rotation) and then has 2 variants per face (mirrored on both axis). So 12 states for them. Skystone will just have 4. It might be nice to use the 4 leftover states to allow fixtures attached to `NEWS` to be rotated by 90 degree, but that is really optional.

That is one of the reason I would recommend moving the decorative stuff into their own mod or package. These things are extremely simple and can be fully covered by the normal minecraft blockstates. So no need to use a full blown `IBakedModel` with rotations and what not. Just stick it somewhere else, maybe have abstract base classes for single textured blocks, log types or model based ones. And  they will not interfere with the more complex ones.

Elix-x commented:
> Also being rotatable is nearly an integral part in 1.8+, as block states have a rotate method (and similar).
> Further it is not decocraft just a convention, that blocks are rotatable.

Yes but no. 90% of mc blocks are still not rotatable. And it's simply illogical to make stone rotatable. Just why?

> Random texture is useless, because that is a hard choice. Either you have it or not. So no possibility to have randomised meteors AND a flat wall.

Let's say we made 4 different textures. Chance that 2 skystones will have same texture is 1/16. 3 - 1/256... So there's extremely low chance that all wall will be the same.

Different mods, huh? We will see (soon).

yueh commented:
> Yes but no. 90% of mc blocks are still not rotatable. And it's simply illogical to make stone rotatable. Just why?

Vanilla is unimportant here. It is an existing feature and that is reason enough.
Also it is nothing complex. 2 Properties, handle the blockstate rotation correctly, a few lines of JSON and done. The only bugs it could cause is something like rotated cw instead of ccw. And pretty much any existing wrench in 1.10 can already rotate it due to a common interface.
The complexity for it is just trivial compared to the rest of AE.

> Let's say we made 4 different textures. Chance that 2 skystones will have same texture is 1/16. 3 - 1/256... So there's extremely low chance that all wall will be the same.

In other words it is impossible to have a wall with just one texture, which is a requirement.

> Different mods, huh? We will see (soon).

It is still interesting not to be common knowledge to put multiple mods into the same .jar.
Say have AE|Core (Current appeng.core, like config files, loading features), AE|Decorative (basic blocks, slabs, stairs), AE|WorldGen (meteor&certus quartz spawning), AE|Tools (simple quartz tools, maybe the grindstone) and AE|ME (everthing related to ME networks). Bundle it together and release it as single .jar. But you can pretty much forget about ever touching the basic ones like decorative or tools after porting. Porting can be more easily done in iterative process when necessary like only releasing the basic ones first, easier to delegate a single part to a person, etc.

Elix-x commented:
> Vanilla is unimportant here. It is an existing feature and that is reason enough.
> Also it is nothing complex. 2 Properties, handle the blockstate rotation correctly, a few lines of JSON and done. The only bugs it could cause is something like rotated cw instead of ccw. And pretty much any existing wrench in 1.10 can already rotate it due to a common interface.
> The complexity for it is just trivial compared to the rest of AE.

But we should not abuse them either.
Flipping torch on an axis - fine ( @shartte you know what to do next).
Rotaing stone - abuse in some way. Let's say that tiniest meteorite left a trail of 5*10 skystones. With 4 different textures, 1 meteorite in 16<sup>5x10-1</sup> (1.0043363e+59) will have all blocks have the same texture...

> In other words it is impossible to have a wall with just one texture, which is a requirement.

Not impossible, but universe will burn out before 10 billion people at 1 rerender a tick will find it.

> It is still interesting not to be common knowledge to put multiple mods into the same .jar.
> Say have AE|Core (Current appeng.core, like config files, loading features), AE|Decorative (basic blocks, slabs, stairs), AE|WorldGen (meteor&certus quartz spawning), AE|Tools (simple quartz tools, maybe the grindstone) and AE|ME (everthing related to ME networks). Bundle it together and release it as single .jar. But you can pretty much forget about ever touching the basic ones like decorative or tools after porting. Porting can be more easily done in iterative process when necessary like only releasing the basic ones first, easier to delegate a single part to a person, etc.

Soon in a good way (i personally agree with modularity, but there might be other opinions).

yueh commented:
> But we should not abuse them either.

I would not call it abuse. Currently every block uses 16 ids, why not make use of them? It is not that it adds any real performance issue or is ridiculous complex to implement. It is just wasting already existing ids, which could be used for some functionality for nearly no costs and there are enough players who appreciate these options.

Elix-x commented:
> I would not call it abuse.

&nbsp;

> Flipping torch on an axis - fine
> Rotating stone - abuse in some way

Yes, it is not hard to implement, but it will take time (which we have, but not much). Though it can come after release and/or when we're done with more complex things.

shartte commented:
> Flipping torch on an axis - fine ( @shartte you know what to do next).

Uuuuuh what? :)
Just to understand correctly... This is multiplying the state space by 2 so the UP direction can be any of the 6 directions, and a new property, let's say FLIPPED=true|false?
So we end up with 12 variations, which will fit into meta.

Elix-x commented:
Yep, new property `FLIPPED` or `MIRRORED` which can take 4th bit (3bits for facing 1 bit for mirror).

yueh commented:
It is not really flipped.

The current fixture can do these things:
- Be attached to another block => 6 states
- The (top) tip of the crystal can either face NE or SW => 2 states (or maybe NW and SE, no idea currently)

So yes, 12 permutations in total, It is also specific to quartz fixtures and not every existing AE block.
There would still be 4 free ids, but that is not enough to cover rotating the crystal in every cardinal direction, which would need 4 states per face => 24 in total. It might be enough to rotate them when attached to top or bottom. But that would be strange. Another idea would be to allow them to be rotated when attached to a cardinal face. But without the mirrored option in this orientation.

Even when time is limited, there is no reason to "pollute" `AEBaseBlock` with properties only related to quartz fixtures. Some contributor will see them and think it is a good idea to reuse them for other cases. Even when they are just wrong because these are already provided by the model baking for more advanced blocks.

Elix-x commented:
More `mirrored` than `flipped`.
Added to todo list for @shartte (in #14).
